File: playwright-test-demo/pages/hovers_page.py
"""
HoversPage -- Page Object for the Hovers page.

Target: https://the-internet.herokuapp.com/hovers
"""

import logging

from pages.base_page import BasePage
from playwright.sync_api import Page

logger = logging.getLogger(__name__)


class HoversPage(BasePage):
    """Page object for /hovers."""

    # -- Selectors --------------------------------------------------------- #
    FIGURES: str = ".figure"
    FIGURE_IMAGES: str = ".figure img"
    FIGURE_CAPTIONS: str = ".figcaption"
    FIGURE_CAPTION_HEADERS: str = ".figcaption h5"
    FIGURE_CAPTION_LINKS: str = ".figcaption a"

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    # -- Actions ----------------------------------------------------------- #

    def open(self) -> "HoversPage":
        """Navigate to the hovers page."""
        self.navigate("/hovers")
        return self

    def hover_over_figure(self, index: int) -> "HoversPage":
        """Hover over a figure by its zero-based index.

        Args:
            index: Zero-based index of the figure to hover over.
        """
        logger.info("Hovering over figure at index %d", index)
        self.page.locator(self.FIGURES).nth(index).hover()
        return self

    # -- State readers ----------------------------------------------------- #

    def get_figure_count(self) -> int:
        """Return the number of figures on the page."""
        return self.page.locator(self.FIGURES).count()

    def is_caption_visible(self, index: int) -> bool:
        """Return whether the caption for a figure is visible.

        Args:
            index: Zero-based index of the figure.

        Returns:
            ``True`` if the caption is visible after hovering.
        """
        return self.page.locator(self.FIGURE_CAPTIONS).nth(index).is_visible()

    def get_caption_text(self, index: int) -> str:
        """Return the caption header text for a figure.

        Args:
            index: Zero-based index of the figure.

        Returns:
            The ``h5`` text inside the caption.
        """
        return self.page.locator(self.FIGURE_CAPTION_HEADERS).nth(index).inner_text().strip()

    def get_profile_link(self, index: int) -> str:
        """Return the profile link href for a figure.

        Args:
            index: Zero-based index of the figure.

        Returns:
            The ``href`` attribute value.
        """
        return self.page.locator(self.FIGURE_CAPTION_LINKS).nth(index).get_attribute("href") or ""

File: selenium-java-demo/src/test/java/pages/HoversPage.java
package pages;

import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;

import java.util.List;

/**
 * Page object for the Hovers page.
 * <p>
 * Target: <a href="https://the-internet.herokuapp.com/hovers">/hovers</a>
 */
public class HoversPage extends BasePage {

    // -- Selectors -------------------------------------------------------- //
    private static final By FIGURES        = By.cssSelector(".figure");
    private static final By FIGURE_CAPTION = By.cssSelector(".figcaption");
    private static final By CAPTION_HEADER = By.tagName("h5");
    private static final By PROFILE_LINK   = By.cssSelector("a[href^='/users/']");
    private static final By PAGE_HEADING   = By.tagName("h3");

    public HoversPage(WebDriver driver) {
        super(driver);
    }

    // -- Actions ---------------------------------------------------------- //

    @Step("Open Hovers page")
    public HoversPage open() {
        navigateTo("/hovers");
        return this;
    }

    @Step("Hover over figure at index: {index}")
    public HoversPage hoverOverFigure(int index) {
        List<WebElement> figures = driver.findElements(FIGURES);
        WebElement figure = figures.get(index);
        new Actions(driver).moveToElement(figure).perform();
        return this;
    }

    // -- State readers ---------------------------------------------------- //

    /**
     * Get the hover info text displayed over a figure after hovering.
     *
     * @param index 0-based figure index
     * @return caption header text (e.g. "name: user1")
     */
    public String getHoverText(int index) {
        List<WebElement> figures = driver.findElements(FIGURES);
        WebElement caption = figures.get(index).findElement(CAPTION_HEADER);
        return caption.getText().trim();
    }

    /**
     * Check if the hover info (figcaption) is visible for a given figure.
     *
     * @param index 0-based figure index
     * @return true if the caption is displayed
     */
    public boolean isHoverInfoVisible(int index) {
        List<WebElement> figures = driver.findElements(FIGURES);
        WebElement caption = figures.get(index).findElement(FIGURE_CAPTION);
        return caption.isDisplayed();
    }

    /**
     * Get the profile link href for a given figure (must be hovered first).
     *
     * @param index 0-based figure index
     * @return href value (e.g. "/users/1")
     */
    public String getProfileLink(int index) {
        List<WebElement> figures = driver.findElements(FIGURES);
        WebElement link = figures.get(index).findElement(PROFILE_LINK);
        return link.getAttribute("href");
    }

    public int getFigureCount() {
        return driver.findElements(FIGURES).size();
    }

    public String getPageHeading() {
        return getText(PAGE_HEADING).trim();
    }
}
